src/logic/function/basic.lean

```diff
@@ -540,34 +540,53 @@ along a function `f : α → β` to a function `β → γ`,
 by using the values of `g` on the range of `f`
 and the values of an auxiliary function `e' : β → γ` elsewhere.
 
-Mostly useful when `f` is injective. -/
+Mostly useful when `f` is injective, more generally when `g.factors_through f`. -/
 def extend (f : α → β) (g : α → γ) (e' : β → γ) : β → γ :=
 λ b, if h : ∃ a, f a = b then g (classical.some h) else e' b
 
+/-- g factors through f : `f a = f b → g a = g b` -/
+def factors_through (g : α → γ) (f : α → β) : Prop :=
+∀ ⦃a b⦄, f a = f b → g a = g b
```

Comment:
I thought it would, but it doesn't  : if \alpha and \gamma are empty,
then f.factors_through g for stupid reasons that there is nothing to prove, 
but you can't construct a function from \beta \to \gamma unless \beta is itself empty.
If \gamma is not empty, one can use function.extend to extend g along f:
`extend f g (const β (inhabited.default : γ))` does the job.
Do you want me to include that ?

```
example (g : α → γ) [inhabited γ]: g.factors_through f ↔ ∃ (e : β → γ), g = e ∘ f :=
begin
  split,
  { intro hf, 
    let e := extend f g (const β (inhabited.default : γ)),
    have he : g = e ∘ f, 
    funext, simp only [comp_app, e], 
    rw extend_apply_of_factors_through g _ hf x, 
    exact ⟨e, he⟩, },
  { intro h, 
    intros a b hf, rw classical.some_spec h, 
    simp only [comp_apply, hf] }
end
```

Reply:
Yes, that example would be good to add (with `[inhabited γ]` replaced by `[nonempty γ]`)

Reply:
Following a request of @fpvandoorn , I added the example.